forge/routes/api/applicationDeviceGroup.js

```diff
@@ -0,0 +1,331 @@
+/**
+ * Application DeviceGroup api routes
+ *
+ * - /api/v1/applications/:applicationId/devicegroups
+ *
+ * @namespace application
+ * @memberof forge.routes.api
+ */
+
+const { registerPermissions } = require('../../lib/permissions')
+const { Roles } = require('../../lib/roles.js')
+
+/**
+ * @param {import('../../forge.js').ForgeApplication} app The application instance
+ */
+module.exports = async function (app) {
+    // ### Routes in this file
+    // GET   /api/v1/applications/:applicationId/devicegroups
+    //       - get a list of devicegroups in this application
+    // POST  /api/v1/applications/:applicationId/devicegroups
+    //       - add a new Device Group to an Application
+    //       > body: { name, [description] }
+    // PUT   /api/v1/applications/:applicationId/devicegroups/:groupId
+    //       - update a device group settings
+    //       > body: { name, [description] }
+    // GET   /api/v1/applications/:applicationId/devicegroups/:groupId
+    //       - get a specific deviceGroup (must be assigned to this application)
+    // PATCH /api/v1/applications/:applicationId/devicegroups/:groupId
+    //       - update Device Group membership
+    //       > OPTION1: body: { add: [deviceIds], remove: [deviceIds] }
+    //       > OPTION2: body: { set: [deviceIds] }
+    // DELETE /api/v1/applications/:applicationId/devicegroups/:groupId
+    //       - delete app owned deviceGroup
+
+    registerPermissions({
+        'application:devicegroup:create': { description: 'Create a device group', role: Roles.Owner },
+        'application:devicegroup:list': { description: 'List device groups', role: Roles.Member },
+        'application:devicegroup:update': { description: 'Update a device group', role: Roles.Owner },
+        'application:devicegroup:delete': { description: 'Delete a device group', role: Roles.Owner },
+        'application:devicegroup:read': { description: 'View a device group', role: Roles.Member },
+        'application:devicegroup:membership:update': { description: 'Update a device group membership', role: Roles.Owner }
+    })
+
+    // pre-handler for all routes in this file
+    app.addHook('preHandler', async (request, reply) => {
+        // get the device group
+        const groupId = request.params.groupId
+        if (groupId) {
+            request.deviceGroup = await app.db.models.DeviceGroup.byId(groupId)
+            if (!request.deviceGroup || request.deviceGroup.ApplicationId !== request.application.id) {
+                reply.code(404).send({ code: 'not_found', error: 'Not Found' })
+            }
+        }
+    })
+
+    /**
+     * Get a list of device groups in an application
+     * @method GET
+     * @name /api/v1/applications/:applicationId/devicegroups
+     * @memberof forge.routes.api.application
+     */
+    app.get('/', {
+        preHandler: app.needsPermission('application:devicegroup:list')
+        // TODO: When mature and ready to be exposed

```

Comment:
This file also appears to have schemas that can be now set!

Reply:
They will be a follow up task, when the API is accepted and being used. Certain things are difficult to foresee and often don't reveal required changes until you use them in real world front facing places.  So did not want to expose APIs that might need updates or changes.